### PR TITLE
Inheritance-based framework detection

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Related to #21761 

Problem: In some functions, we detect the framework of the model class through its name (e.g. if it starts with `TF`). This is a quirk of our library, and users might run into issues due to this hidden behavior. For instance, in the issue linked above, a user created a child class of a TensorFlow model whose name did not start with `TF`, running into exceptions.

Solution: Inheritance-based framework detection :)

### CODE REVIEW

File: src/transformers/utils/generic.py [hunk 1]
```diff
@@ -366,13 +366,14 @@ def can_return_loss(model_class):
...
-        signature = inspect.signature(model_class.__call__)
+    base_classes = str(inspect.getmro(model_class))
+
+    if "keras.engine.training.Model" in base_classes:
+        signature = inspect.signature(model_class.call)  # TensorFlow models
+    elif "torch.nn.modules.module.Module" in base_classes:
+        signature = inspect.signature(model_class.forward)  # PyTorch models
     else:
-        signature = inspect.signature(model_class.forward)
+        signature = inspect.signature(model_class.__call__)  # Flax models
```
Comment (ArthurZucker) [inline src/transformers/utils/generic.py:376]: nice catch

**Review by ArthurZucker — APPROVED** [review]
Looks good to me. We can also add a small test with the minimal reproduction script I shared on the issue 😉

**Comment by gante** [comment]
Now with a test 👍

File: tests/utils/test_file_utils.py [hunk 2]
```diff
@@ -85,29 +95,39 @@ def test_context_managers_two_context(self, mock_stdout):
         # The output should be wrapped with an English and French welcome and goodbye
         self.assertEqual(mock_stdout.getvalue(), "Bonjour!\nWelcome!\nTransformers are awesome!\nBye!\nAu revoir!\n")
 
-    def test_find_labels(self):
```
Comment (gante) [inline tests/utils/test_file_utils.py:88]: In this test:
- moved imports to the top
- separated it into 3 tests (one per framework)
- added a check to confirm that it works regardless of the class name

**Review by sgugger — APPROVED** [review]
Nice new tests, thanks a lot for working on this!

**Review by Rocketknight1 — APPROVED** [review]
Good clean solution, with one question! Is `keras.engine.training.Model` always one of the parent classes for TF models in all versions of TF? I suspect it's true in more recent versions after they split Keras out from core TF a bit more, but in e.g. 2.5 or something is that still the right base class?

**Comment by gante** [comment]
@Rocketknight1 good point! I have no clue if it holds for TF2.5 -- going to double check

**Comment by gante** [comment]
@Rocketknight1 hah, the class is different but it still works! In TF2.4, we have `<class 'tensorflow.python.keras.engine.training.Model'>` in the inheritance tree, so `"keras.engine.training.Mode"` is still a match :D

No change is needed, but thanks for the shoutout 👌

**Comment by nidhishgajjar** [comment]
> **Orb Code Review** (powered by GLM 5.1 on [Orb Cloud](https://orbcloud.dev))

Nice improvement — using inheritance-based framework detection instead of name-prefix matching is more robust and handles custom subclasses correctly. The test coverage is excellent, especially the `DummyModel` subclass tests that verify the fix works for non-standard class names.

---

### 🟡 Should Fix

**1. Duplicated framework detection logic**

The same MRO-based string check is repeated in 4 places across 2 files:
```python
framework = "tf" if "keras.engine.training.Model" in str(inspect.getmro(model.__class__)) else "pt"
```
Consider extracting this into a utility function like `detect_framework(model_class)` in `src/transformers/utils/generic.py` to avoid duplication and make future maintenance easier.

**2. String-based MRO matching is fragile**

Using `"keras.engine.training.Model" in str(inspect.getmro(...))` works but could be brittle:
- If Keras changes internal class paths (e.g., `tf.keras.Model` vs standalone `keras.Model`)
- If a user's custom class path happens to contain this substring

A more robust approach would be to use `issubclass` with the actual class:
```python
import keras
if issubclass(model_class, keras.Model):
    ...
```
This would need proper import guards but is more future-proof.

---

### ℹ️ Minor / Observations

**3. Mixed detection strategies in `find_labels`**

`find_labels` now uses MRO for framework detection but still uses name-based detection for the QA special case (`if "QuestionAnswering" in model_name`). This is fine since QA model names do consistently contain "QuestionAnswering", but it's worth noting the mixed strategy.

**4. Good test restructuring**

The test changes are well done — splitting the monolithic `test_find_labels` into separate `@require_torch`, `@require_tf`, `@require_flax` methods is cleaner and properly handles environments where some frameworks aren't installed.

---

### Summary
This is a solid improvement that fixes a real limitation of name-prefix-based framework detection. The core logic is correct and well-tested. Consider extracting the duplicated detection logic into a utility function and using class-based `issubclass` checks for more robust detection.

### Assessment: comment

_This review was created by an AI assistant (OpenHands) on behalf of Orb Cloud._

**Pull request merged**